### DISCUSSION TITLE
Using Typst instead of LaTeX?

### DISCUSSION THREAD

**Discussion by sinaatalay** [rendercv/rendercv#233] [Ideas]
[Typst](https://typst.app) is an alternative to $\LaTeX$. It has its own markup language, but in the end, it takes a text file and gives you a PDF.

Typst is written in Rust. It's very fast and can run almost anywhere (on the web, mobile phones, etc.)

RenderCV's idea would be the same: RenderCV will create a Typst source file with the templates from the YAML input file and call Typst to render it into a PDF.

What would you think of switching to Typst? It may help the cross-platform RenderCV app we are developing right now.

**Comment by mkbhuiyan96** [comment]
Would the output look identical?

**Reply by sinaatalay** [reply]
We have already switched to Typst with v2.0. Except `moderncv`, I think they are identical.

**Comment by guilhermefgs** [comment]
Latex is a well known resource in the academic community. I was used to manipulating my custom themes to design my CVs and Cover Letters with Latex code. 

Adding typst to support rendercv app is removing the flexibility of using latex, which allow designing custom themes easily. One example is the new Preamble of moderncv that has now 450+ lines of code.

I was using a custom theme based on moderncv and understand that the current template tries to mimic the latex package. ModernCV package had cover letter features that I was using to generate such documents, matching my CV style.

In conclusion, I think this change might be good for the app, but not for the users that want to customize CVs with their latex knowledge.

**Comment by sinaatalay** [comment]
Thank you for sharing your thoughts @guilhermefgs. Here's why we made this decision:

We have limited time and resources to develop RenderCV, so we carefully prioritize our efforts to create the best possible user experience with the least amount of energy.

1. LaTeX was draining our energy due to the following reasons:
    1. It was very hard to incorporate LaTeX into RenderCV without asking users to download and install LaTeX. We wanted RenderCV to work out of the box with `pip install rendercv`; therefore, we had to ship a LaTeX distribution with RenderCV. However, there is a file size limit for PyPI packages, so we had to write custom scripts to minimize TinyTeX to 50 MB for RenderCV. It was really tough to add new packages and fonts after that.
    2. We couldn't support Chinese, Korean, Japanese, Greek, etc., with LaTeX. We basically couldn't upload all the required files to PyPI.
    3. Maintaining the LaTeX templates was extremely difficult. We had many ideas for design options that we couldn't achieve because of LaTeX's inherent difficulties.

2. The user experience of LaTeX wasn't great as well:
    1. It was very slow, and it couldn't leverage parallel computing. Each CV took ~2 seconds to render.
    2. If you are not familiar with LaTeX already, playing with LaTeX templates is really hard.
    3. LaTeX is not cross-platform. You cannot run it reliably with web assembly. It's hard to run them on mobile devices, too. So, if you wanted to build an app with RenderCV, you needed a backend server running LaTeX.

We thought about this a lot and realized there was already a solution to all of these problems: Typst. It's a new and emerging technology that is in the ascendancy over LaTeX.

How does Typst address the problems above?

1.  Point 1
    1. It's extremely easy to incorporate Typst into RenderCV because Python bindings already exist! See the [`typst`](https://pypi.org/project/typst/) package.
    2. Typst supports Chinese, Korean, Japanese, Greek, etc., out of the box, and it's just 10 MB.
    3. Typst language is much more powerful than LaTeX, and maintaining the templates is much easier. You pointed out that Preamble.j2.typ of the `moderncv` theme has 450+ lines of code. That's because all themes share the same templates, and 450+ lines of code govern all the templates, which we could never achieved with LaTeX.
2. Point 2
    1. Typst is extremely fast. It allowed RenderCV to work in real time. Check out this [demo video](https://www.youtube.com/watch?v=dZ17UrXBmWw).
    2. Typst language is really easy and intuitive to use.
    3. Typst is cross-platform. It will allow us to run the app completely on the client side.

Typst was the obvious choice. We could have supported both LaTeX and Typst in v2, but we decided not to spend our energy doing both; instead, we decided to focus on one and make it great.

I understand LaTeX is very popular. However, our focus is on the user experience instead of selecting popular technologies for broader appeal.

Finally, RenderCV v1.18 (`pip install rendercv==1.18`), the latest version that uses LaTeX will always be available. While it won’t receive further maintenance, if one finds value in it, one could fork and maintain it independently.

**Reply by guilhermefgs** [reply]
Thank you for the detailed explanation.

After some hours updating my CV to the newest version, I observed that the new and extensive configuration added to the templates supports most of the changes I would want to apply to the default themes, making the custom themes less necessary.

Because of the extensive configuration, the typst codes are in my opinion not easy and intuitive to change. But as you mentioned, the advantages over latex makes it a necessary trade. Let's welcome typst and embrace the future :smile:
